address comments on the general usability

liu-ziyang opened the pull request:
Updated UI for the CLI to be more usable.
1. splited validation result to different sections where the errors are at the end, and individual PASS/FAIL check at the beginning
2. updated syntax to `npd validate <package name>` to run on a specific package
3. separated out modules to npd validate and npd discover for different functionality
4. added coloring of the result rendering
5. included a CLI_usage.md to demonstrate usages

see example below to get more intuitive idea of what it looks like now:

<img width="799" alt="Screen Shot 2021-03-30 at 12 46 57 PM" src="https://user-images.githubusercontent.com/41968256/113047346-0ca09e80-9156-11eb-876a-aa6702363afa.png">


npd discover (without and with verbose flag)
<img width="660" alt="Screen Shot 2021-03-30 at 12 17 05 PM" src="https://user-images.githubusercontent.com/41968256/113047164-d400c500-9155-11eb-84f6-46b07735a98a.png">

Commit:
address comments on the general usability

Commit:
minor fix

Commit:
update pre run readme

Commit:
update grammar

Commit:
added coloring of validation result, refactored the code to be more modular, and streamlined checks

neuromusic commented:
thanks for addressing some of the the comments in the doc, @ziyangczi ! I'm going to dive in with a review here focusing on the CLI usage.

neuromusic commented:
IDK if its b/c this is marked as WIP or b/c I'm not a reviewer or b/c I'm not an admin, but I can't "review" the PR.

I'll give my feedback below.

neuromusic commented:
`documentation/CLI_usage.md`

> ### Description
> napari plugin devtools scans the current python environment to find out details about napari plugin information.

its good that the documentation makes it explicit what the tool does, but I fear that this is going to be at best confusing/unintuitive for users and at worst provide inaccurate information.

as a plugin dev, I want to know about the validity of my plugin, which exists as code in a folder on my computer. this tool requires me to first install the plugin and this tool into an environment, then run the tool from the command line. Two thoughts here:

1. I can't think of any other CLI tool I've used that works this way. Do you have any good examples of this pattern?
2. This assumes that the environment where the plugin and napari-plugin-devtools are installed are the "active" environment at the command line. I've worked with enough developers who have gotten PATH wired up in weird ways such that CLI command do not always correspond to what conda or venv considers to the the "active" environment. So there's a very large risk that when the user calls `npd` from the command line it does not call `npd` in the "active" environment, but rather calls `npd` in another Python environment on their system, which is very hard to debug because of (1)

I'm curious if there are solutions here that can avoid this confusion by letting the user point to a specific repository, folder, or built package? If the validation logic requires installation of the plugin, can `npd` build (and tear down) its own environment? e.g. something like how `tox` works?

neuromusic commented:
> ### Usage
> ```
> npd validate [Options] <package> [<packages>...]
> ```

Requiring a package makes sense to me. 

It does not make sense to me to allow multiple packages to be validated. Can you expand on the motivation and use case for testing multiple plugins with one command?

neuromusic commented:
> validate single package
> ```
> $ pip install napari-svg
> $ npd validate napari-svg

While this does demonstrate the usage, a demo using `pip install -e ./napari_svg` would be closer to a real-world example.

> ```
> ----------------------------------------------------------------
> Scanning current python environment: /Library/Frameworks/Python.framework/Versions/3.8
> ```

I like the clarity around which environment is being scanned given the risk of ambiguity here.

> ```
> ----------------------------------------------------------------
> Validating package napari-svg
> * Classifier check: FAILED
> * Entrypoint check - syntax: PASSED
> * Entrypoint check - plugin registration: PASSED
> * Entrypoint check - registered modules exist: PASSED
> * Hooks registration check - svg: PASSED
> ```

I like the explicit list of checks with PASS/FAIL for each.

> ```
> ----------------------------------------------------------------
> Error: 'Framework :: napari' does not exist for napari-svg's 12 known classifier(s)
> ----------------------------------------------------------------
> ```

Adding more detail on the FAILED item: 💯

neuromusic commented:
> validate multiple packages with verbose output:

This example drives home that I don't like combining the validation of multiple packages into one CLI call. We should keep two packages two separate CLI calls.

> ```
> ----------------------------------------------------------------
> Validating package napari-svg
> * Classifier check: FAILED
> * Entrypoint check - syntax: PASSED
> * Entrypoint check - plugin registration: PASSED
>         Plugins registerd under napari-svg: ['svg']
> * Entrypoint check - registered modules exist: PASSED
>         Modules found: ['napari_svg']
> * Hooks registration check - svg: PASSED
>         Found 6 registered writer hook(s) for svg:
>         - napari_get_writer
>         - napari_write_image
>         - napari_write_labels
>         - napari_write_points
>         - napari_write_shapes
>         - napari_write_vectors
> ```

I don't care for interspersing the "verbose" feedback with the checklist of PASS/FAIL. I think it will be more clear if the `--verbose` flag instead added a "verbose" section above the checklist.


```
----------------------------------------------------------------
Checking for `Framework :: napari` classifier... 
Classifier was not found.
Checking for `napari.plugin` entrypoint... 
Entrypoint found.
Checking syntax of the `napari.plugin` entrypoint... 
Syntax looks good.
Checking for plugins registered... 
Found "svg" registered in the `napari-svg` package.
Checking that the entrypoint's target modules exist... 
Found `napari_svg` for the "svg" plugin
Checking for hook implementations in each target module...
Found 6 registered writer hook(s) for "svg": napari_get_writer, napari_write_image, napari_write_labels, napari_write_points, napari_write_shapes, napari_write_vectors
----------------------------------------------------------------
Validation summary for napari-svg
* Classifier check: FAILED
* Entrypoint check - syntax: PASSED
* Entrypoint check - plugin registration: PASSED
* Entrypoint check - registered modules exist: PASSED
* Hooks registration check - svg: PASSED
```

justinelarsen reviewed (commented):
I love the updates! Especially the CLI usage documentation which is very helpful to know exactly what commands are available, how to use them and what to expect. Colors for 'pass' and 'fail' are also a nice touch. Great work @ziyangczi !

justinelarsen commented on documentation/CLI_usage.md:
````diff
@@ -0,0 +1,195 @@
+```
+
+### Commands
+|Command        |Description                                          |
````
Love the table and new format - easy to read and understand what commands are available.
Also like that you split out the discover command from the validate.

neuromusic replied:
+1

justinelarsen commented on documentation/CLI_usage.md:
```diff
@@ -0,0 +1,195 @@
+|Command        |Description                                          |
+|---------------|-----------------------------------------------------|
+|validate       |validate the given package is correctly setup        |
+|discover       |discover all plugins under current python environment|
```
"list" or "ls" may be shorter commands. Some things to consider are if the user would be confused about what they are listing though. I just wanted to put other options out there to see what others think.

justinelarsen commented on documentation/CLI_usage.md:
```diff
@@ -0,0 +1,195 @@
+$ pip install napari-svg
+$ npd validate napari-svg
+----------------------------------------------------------------
+Scanning current python environment: /Library/Frameworks/Python.framework/Versions/3.8
```
I prefer to scan my current directory or be able to specify a directory instead of my current python environment so that I can validate my plugin as I'm developing it and before I publish to Pypi. But if we are going to keep this method, I really like that you've added this output so that I know where the tool is scanning; it's very helpful.

justinelarsen commented on documentation/CLI_usage.md:
```diff
@@ -0,0 +1,195 @@
+----------------------------------------------------------------
+Scanning current python environment: /Library/Frameworks/Python.framework/Versions/3.8
+----------------------------------------------------------------
+Validating package napari-svg
```
I really like the summary here with the different checks and clear output of pass/fail.

justinelarsen commented on documentation/CLI_usage.md:
```diff
@@ -0,0 +1,195 @@
+* Entrypoint check - registered modules exist: PASSED
+* Hooks registration check - svg: PASSED
+----------------------------------------------------------------
+Error: 'Framework :: napari' does not exist for napari-svg's 12 known classifier(s)
```
Error location is good, but I still could be confused on what to do to resolve this. Part of the error message is a little confusing with "12 known classifier(s)". Could it also point me to a line in the code or even just a file name where I should add this classifier?

liu-ziyang replied:
included tutorial page for similar error messages

justinelarsen commented on README.md:
```diff
@@ -18,35 +18,50 @@ these checks verify that the plugin is available for users of napari
-The validation would run on built packages under dist folder to check if they are annotated
-properly with framework classifier, and validate hooks are properly annotated that they can be found
-by napari.
+Before running `npd validate`, make sure you have prepared the repo for validation:
```
Can we also offer the option to scan my current directory or be able to specify a directory instead of my current python environment so that I can validate my plugin as I'm developing it and before I publish to Pypi?

justinelarsen commented on README.md:
```diff
@@ -18,35 +18,50 @@ these checks verify that the plugin is available for users of napari
-command line interface and the other is a pytest fixture.
+command line interface, and the other is a pytest fixture.
 
 ### Command Line Interface (CLI) Usage
```
Consider adding a link to the CLI usage documentation (documentation/CLI_usage.md) which is super helpful.

justinelarsen commented on documentation/CLI_usage.md:
```diff
@@ -0,0 +1,195 @@
+* Entrypoint check - registered modules exist: PASSED
+        Modules found: ['napari_svg']
+* Hooks registration check - svg: PASSED
+        Found 6 registered writer hook(s) for svg:
```
I like that we're listing out the hooks that were found now in verbose mode. This is helpful to see all the hooks that are registered.

justinelarsen commented on documentation/CLI_usage.md:
```diff
@@ -0,0 +1,195 @@
+----------------------------------------------------------------
+Scanning current python environment: /Library/Frameworks/Python.framework/Versions/3.8
+----------------------------------------------------------------
+Found 1 registered function hook(s) for napari-demo:
```
I would like to see a list of discovered plugins here, not including the hooks. If it's useful to also include hooks here, I would show them under the list or put them under a --verbose output. Also having the plugin name on a line above each "Found...hook(s) for..." would be easier to read.

liu-ziyang replied:
moved to verbose to match with validation consistently

justinelarsen commented on napari_plugin_devtools/__main__.py:
```diff
@@ -2,111 +2,231 @@
+
+from .validation import Options, list_hook_implementations
+
+PASSED = colored('PASSED', 'green')
```
Love the colors! makes it easier to read quickly

neuromusic replied:
yes! love this

justinelarsen commented on README.md:
```diff
@@ -18,35 +18,50 @@ these checks verify that the plugin is available for users of napari
 
-`npd validate -v|--verbose` enable verbose mode, gives slightly more information on the underlying 
-findings of validation process.
+Output from `npd validate` has multiple sections, where each section is a separate validation where the header
```
I like that you explain what each section is for so I know what will be validated. I think the order here is actually different than the output though which could be confusing. It would be good to have these checks in the same order as the output when I run  the validate command.

liu-ziyang replied:
pointed to the usage doc to avoid duplicate/inconsistent documentation

Commit:
- constraint support to single package
- updated error message to include tutorial sections
- updated verbose message to show at the end
- moved hook display to verbose mode consistently
- updated readme to point to usage doc for CLI

Commit:
update documentation to reflect local installation of plugin and error messages

liu-ziyang commented:
> > ### Usage
> > ```
> > npd validate [Options] <package> [<packages>...]
> > ```
> 
> Requiring a package makes sense to me.
> 
> It does not make sense to me to allow multiple packages to be validated. Can you expand on the motivation and use case for testing multiple plugins with one command?

as discussed in meeting, going to constraint this to single package

liu-ziyang commented:
> I don't care for interspersing the "verbose" feedback with the checklist of PASS/FAIL. I think it will be more clear if the `--verbose` flag instead added a "verbose" section above the checklist.

moved verbose output to a separate section

liu-ziyang commented:
Answering some of the feedback around local installation: there is some challenge to create a temporary python environment, installing the folder in it, and switch to that environment to run checks in code, unfortunately for now we would have to rely on the user to install the local code via pip, I have updated the instruction to be clear on this

justinelarsen commented:
> Answering some of the feedback around local installation: there is some challenge to create a temporary python environment, installing the folder in it, and switch to that environment to run checks in code, unfortunately for now we would have to rely on the user to install the local code via pip, I have updated the instruction to be clear on this

I think I'm missing something. Can you provide more info on why we need the installation? If we're just scanning files to look for certain contents to check for classifiers, hooks, etc., can't we just run a find command on a directory? Why do we need a python environment for that?

liu-ziyang commented:
> > Answering some of the feedback around local installation: there is some challenge to create a temporary python environment, installing the folder in it, and switch to that environment to run checks in code, unfortunately for now we would have to rely on the user to install the local code via pip, I have updated the instruction to be clear on this
> 
> I think I'm missing something. Can you provide more info on why we need the installation? If we're just scanning files to look for certain contents to check for classifiers, hooks, etc., can't we just run a find command on a directory? Why do we need a python environment for that?

Right now we do not scan the files for content, instead, we used the same way as to how napari scans for installations in the python environment. Though switching to the file scan solves this particular problem, it also introduced the problem of possible disconnection between our validation and the native napari registration, therefore could have more bugs introduced along the way. 

And to work around that there is an alternative of creating a temporary environment so we can install the folder and still be able to scan the native way as napari did. which also requires some effort

justinelarsen commented:
> > > Answering some of the feedback around local installation: there is some challenge to create a temporary python environment, installing the folder in it, and switch to that environment to run checks in code, unfortunately for now we would have to rely on the user to install the local code via pip, I have updated the instruction to be clear on this
> > 
> > 
> > I think I'm missing something. Can you provide more info on why we need the installation? If we're just scanning files to look for certain contents to check for classifiers, hooks, etc., can't we just run a find command on a directory? Why do we need a python environment for that?
> 
> Right now we do not scan the files for content, instead, we used the same way as to how napari scans for installations in the python environment. Though switching to the file scan solves this particular problem, it also introduced the problem of possible disconnection between our validation and the native napari registration, therefore could have more bugs introduced along the way.
> 
> And to work around that there is an alternative of creating a temporary environment so we can install the folder and still be able to scan the native way as napari did. which also requires some effort

Ok. I see now. Thanks for clarifying! For the temp environment method, would you still need to package up the plugin within the new environment?

liu-ziyang commented:
> Ok. I see now. Thanks for clarifying! For the temp environment method, would you still need to package up the plugin within the new environment?

nope, we have got rid of packaging in our validation flow completely, only local installation is required, for example from the local folder of plugin `pip install -e .` (this works both for our current flow and the alternative approach of having a temp env, just that the installation would happen in the new vitural env instead of the current active one) and then run validation command. Also added this to the usage readme to reflect the usage better, credit goes to Justin's suggestion

Commit:
update cli usage

sofroniewn reviewed (approved):
As was discussed in our meetings we're going to refocus some of the work here to ensure less duplication of things within napari, and napari hub.

I'm going to merge now as I think this is better than what is there now.

Sounds like the next iteration might be a napari hub focused CLI tool that could build off of this work, but be in a different dedicated repo.
